### PR TITLE
Bugfix for issue 14333 (convolve_models giving strange results)

### LINKED ISSUE #14333: convolve_models giving strange results
### Description
Depending on my x value input sometimes, the y value of my convolved functions will give strange results.

### Expected behavior
I expect the value of the convolved models not to depend on input x, since I've defined analytic functions (astropy.models) inputs to the convolve_models function.

### Steps to Reproduce

```python
import numpy as np
from astropy.modeling import models, fitting
from astropy.convolution import Gaussian1DKernel, convolve, convolve_models
import matplotlib.pyplot as plt

gauss = models.Gaussian1D(.2,0,.2)
box = models.Box1D(.2,0,2)
combined = convolve_models(box, gauss)

x1 = np.linspace(-20,30,100000)
plt.plot(x1, box(x1)+0.05) #vertical offset to make them easier to see
plt.plot(x1, gauss(x1)+0.1)#vertical offset to make them easier to see

plt.plot(x1, combined(x1))

plt.xlim(-30,30)
plt.show()


x2 = np.linspace(-10,10,100000)
plt.plot(x2, box(x2)+0.05) #vertical offset to make them easier to see
plt.plot(x2, gauss(x2)+0.1)#vertical offset to make them easier to see

plt.plot(x2, combined(x2))

plt.xlim(-30,30)
plt.show()

```

### PR DESCRIPTION
**Pull request by subagonsouth**
### Description

This pull request is to address bug reported in issue #14333. The root of this bug
is that the implementation of the astropy.convolution.convolve.convolve_models() function
simply convolves the output of the two input models. I found that the correct behavior
was implemented in the astropy.convolution.convolve.convolve_models_fft() function
and a description of how that function fixes the issue was buried in the docstring
for astropy.modeling.convolution.Convolution class.

A couple of notes:
1. The astropy.modeling.convolution.Convolution class is sensitive to the input
    models as well as the bounding box, resolution, and normalize_kernel argument.
    I tried to write clear documentation to that affect, but am open to improvements.
2. I tried to reference the astropy.modeling.convolution.Convolution class in the documentation
    but it is not found in the API documentation. I could not figure out the fix for this.
3. A more complex example could be added to the docs.modeling.compound-models.rst file
    with an example of using a Convolution combined model with a fitter to try to illustrate how
    to choose appropriate inputs.
4. Now the astropy.convolution.convolve.convolve_models_fft() function is duplicated by the
    functionality of astropy.convolution.convolve.convolve_models_fft(). Should convolve_models_fft()
    raise a Deprecation warning of some sort and plan to be removed in a future release?

Fixes #14333 

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/convolution/convolve.py
Comment (pllim) [inline astropy/convolution/convolve.py:1005]: This is API change, so maybe a second change log entry to explain this change?

**Review by pllim — COMMENTED** [review]
Thanks!

Warnings in RTD log are related and need to be addressed.

Also in Sphinx RST, I think you need a whitespace after colon before listing starts, and there is no need to indent the numbered list w.r.t. parent text.

**Comment by pllim** [comment]
@saimn , if this PR goes forward, you think the API change is ok for 6.1 or should this wait till 7.0?

**Comment by subagonsouth** [comment]
> Thanks!
> 
> Warnings in RTD log are related and need to be addressed.
> 
> Also in Sphinx RST, I think you need a whitespace after colon before listing starts, and there is no need to indent the numbered list w.r.t. parent text.

I will address some of the warning and fix the list issue. As noted in my description, I am not sure why there is no API documentation being generated for the astropy.modeling.convolution module. I believe this is the reason why my references to astropy.modeling.convolution.Convolution are failing. Any help/guidance would be appreciated.

**Comment by pllim** [comment]
> I am not sure why there is no API documentation being generated for the astropy.modeling.convolution module. 

Hopefully `convolution` maintainers (@larrybradley @adonath @keflavich) can advise here. 🤞

File: astropy/convolution/convolve.py
Comment (WilliamJamieson) [reply astropy/convolution/convolve.py:1005]: In order to preserve backwards compatibility, with the API this probably needs to actually be:

```python
def convolve_models(
    model, kernel, mode="convolve_fft", bounding_box=None, resolution=None, cache=True, **kwargs
)
```
where the `None` values map to  assumed defaults. It might be reasonable to issue a warning if they are not specified.

This way no current code which positionally passes the `mode` keyword and/or fails to define the `bounding_box` or `resolution` keywords will be immediately broken.

**Comment by adonath** [comment]
> Hopefully convolution maintainers (@larrybradley @adonath @keflavich) can advise here. 🤞

As far as I can tell the `Convolution` object is not part of the public namespace. It is not imported into the `astropy.modeling` namespace in `astropy.modeling.__init__.py`, so this is expected.The question is whether it should be public...?

**Comment by subagonsouth** [comment]
@adonath 
> The question is whether it should be public...?

I will certainly leave the final decision up to the maintainers. Here are some considerations...

1. Part of what I wanted to do in this PR was make the documentation in the Notes section of the `Convolution` class accessible to users so that they can understand the challenge of providing a function that combines models through convolution and how that has been solved. I was trying to avoid duplicating that documentation while providing a bit more context.
2. When considering the question of should the `Convolution` class be made public, I considered the idea of moving the functionality from `convolve_models()` into that class and making it public. I would not advocate for this because I think that the `convolve_models` function provides an interface that is more similar to all of the other functions for combining models.
3. I don't see any use in making `Convolution` public without making it a bit more general. If you think about it, there is nothing in it that is necessarily specific to convolution. Instead, it provides functionality to combine models using an operator that is sensitive to the domain input to the models in a way that removes that sensitivity to the input domain. I can't think of any other such operators, but they may exist.
4. It occurs to me that it may be prudent to move the `convolve_models()` function from the `convolution.convolve` module into the `modeling.convolution` module. That organization makes more sense to me.

File: astropy/convolution/convolve.py
Comment (neutrinoceros) [reply astropy/convolution/convolve.py:1005]: Since `resolution`, `cache`, and `bounding_box` have never been allowed as positional arguments, an ever more conservative signature would be
```python
def convolve_models(
    model, kernel, mode="convolve_fft", *, bounding_box=None, resolution=None, cache=True, **kwargs
)
```

Comment (subagonsouth) [reply astropy/convolution/convolve.py:1005]: @WilliamJamieson @neutrinoceros 
I will make the recommended change to the `convolve_models()` signature, however I will point out that this only preserves backward compatibility in the API. The functionality has changed such that any calls to the function will undoubtedly need to be updated, adding the new keywords, in order to restore the previous behavior. In fact, making the API backward compatible may obfuscate the issue because users are likely to get a `ValueError: One of the requested xi is out of bounds` error when the underlying interpolation fails because the bounding box has not been properly defined.

Comment (subagonsouth) [reply astropy/convolution/convolve.py]: I have made the requested change, but have to say that I don't like it. It seems buggy to implement default values for `bounding_box` and `resolution` that may not match the dimensionality of the input model and kernel. Maybe I should just rollback the changes to the `convolve_models` function and improve the documentation to note that that function does not behave like one would expect but instead just convolves the output of `model` and `kernel`.

Comment (pllim) [reply astropy/convolution/convolve.py]: > Maybe I should just rollback the changes

If you are up for it, does not hurt to have a competing PR with just the doc changes, and see which one gets merged. If you do that, you can specify that the new PR would close this PR, to be clear. Thanks!

Comment (neutrinoceros) [reply astropy/convolution/convolve.py]: I cannot unresolved this thread but I think it was prematurely closed. I'll open a other one so I'm sure it's visible.

File: astropy/convolution/convolve.py
Comment (neutrinoceros) [inline astropy/convolution/convolve.py:1089]: In my opinion this list comprehension cannot be clever *and* easy to read, and I'd favour readability, so I suggest making it dumb :)
```suggestion
        none_kws_list = []
        if bounding_box is None:
            none_kws_list.append("bounding_box")
        if resolution is None:
            none_kws_list.append("resolution")
        none_kws = " and ".join(none_kws_list)
```

File: astropy/convolution/convolve.py
Comment (neutrinoceros) [inline astropy/convolution/convolve.py:1092]: An additional implicit breaking change here is that this path requires `scipy`, which isn't a required dependency to `astropy` (there's even a 2 years old issue proposing to change that #12448). Again, this could be okay to do in a major release (7.0 is the next one), but it would be a lot more gentle to users if we could warn about it ahead of time, and keep the code working as it does now until then. You'd need to check at runtime whether `scipy` is installed (use `astropy.utils.compat.HAS_SCIPY`) and emit about it being required in the future if it's missing.

**Review by neutrinoceros — CHANGES_REQUESTED** [review]
I think the change in itself is good and desirable, but since breaking changes are needed, we ought to try and make the transition a smooth as possible for users.
The way I see it, the new behaviour shouldn't become the default just yet, and only be used on the condition that user code explicitly pass non-`None` values for `bounding_box` or `resolution` when calling `convolve_models`. There are a handful case we need to take care of (and test):
- existing code that doesn't pass  `bounding_box` or `resolution` should continue to produce identical results for now, but receive a `FutureWarning`. The warning should also mention that scipy will be required by this function in the future (this can be done regardless of wether it is installed already or not).
- if `bounding_box` or `resolution` are received, switch to the new behaviour. In particular, raise a clear error message if `scipy` isn't installed.

I should also note, in case this isn't obvious to everyone (it wasn't to me), that `bounding_box` and `resolution` are not currently recognised keyword arguments to this function, so it's safe to assume that old code isn't already using them.

File: astropy/convolution/convolve.py
Comment (pllim) [reply astropy/convolution/convolve.py]: I unresolved it for you.

**Comment by neutrinoceros** [comment]
@subagonsouth Would you like to continue this work or would you like me to step up and finish it ?

**Comment by subagonsouth** [comment]
@neutrinoceros 
> @subagonsouth Would you like to continue this work or would you like me to step up and finish it ?

Due to some major life changes, I don't anticipate being able to finish this up anytime soon. If you would like to do so, that is fine by me. Thanks!

**Comment by neutrinoceros** [comment]
Thanks for your reply ! I'll circle back to it eventually, possibly not before September, and I'll open a new PR building on top of your commits so you won't need to give me push rights to your fork.

**Comment by neutrinoceros** [comment]
Just a quick update: I haven't forgotten this PR but I will prioritize other projects over the month of october, and plan to come back to this after astropy 7.0.0's feature freeze. Thanks for your understanding.

**Pull request closed without merging**
